consul/state/prepared_query.go

```diff
@@ -16,22 +17,54 @@ func isUUID(str string) bool {
 	return validUUID.MatchString(str)
 }
 
+// queryWrapper is an internal structure that is used to store a query alongside
+// its compiled template, which can be nil.
+type queryWrapper struct {
+	*structs.PreparedQuery
+	ct *prepared_query.CompiledTemplate
+}
+
+// toPreparedQuery unwraps the internal form of a prepared query and returns
+// the regular struct.
+func toPreparedQuery(wrapped interface{}) *structs.PreparedQuery {
+	if wrapped == nil {
+		return nil
+	}
+	return wrapped.(*queryWrapper).PreparedQuery
+}
+
 // PreparedQueries is used to pull all the prepared queries from the snapshot.
-func (s *StateSnapshot) PreparedQueries() (memdb.ResultIterator, error) {
-	iter, err := s.tx.Get("prepared-queries", "id")
+func (s *StateSnapshot) PreparedQueries() (structs.PreparedQueries, error) {
+	queries, err := s.tx.Get("prepared-queries", "id")
 	if err != nil {
 		return nil, err
 	}
-	return iter, nil
+
+	var ret structs.PreparedQueries
+	for wrapped := queries.Next(); wrapped != nil; wrapped = queries.Next() {
+		ret = append(ret, toPreparedQuery(wrapped))
```

Comment:
Any reason to not just `wrapped.(*queryWrapper).PreparedQuery`? You already handle the nil check before we get here it looks like.